File: DSA-loan-management/dsa-mgmt-be/app/api/routers/customers.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import List, Optional

from app.db.session import SessionLocal
from app.models.customer import Customer
from app.models.agent import Agent

router = APIRouter()


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


class CustomerCreate(BaseModel):
    name: str
    email: str
    mobile: str


class CustomerUpdate(BaseModel):
    name: str
    email: str
    mobile: str


class AssignAgentPayload(BaseModel):
    agentId: Optional[int] = None


def _serialize(c: Customer) -> dict:
    return {
        "id": c.id,
        "name": c.name,
        "email": c.email,
        "mobile": c.mobile,
        "uniqueCustomerId": c.uniqueCustomerId,
        "agentId": c.agentId,
        "agentName": c.agent.name if c.agent else None,
        "agentPhoto": c.agent.photo if c.agent else None,
        "status": c.status,
        "isActive": c.isActive,
    }


@router.get("/")
def list_customers(agent_id: Optional[int] = None, include_inactive: bool = False, db: Session = Depends(get_db)):
    query = db.query(Customer)
    if not include_inactive:
        query = query.filter(Customer.isActive == True)
    if agent_id is not None:
        query = query.filter(Customer.agentId == agent_id)
    customers = query.all()
    return [_serialize(c) for c in customers]


@router.get("/{customer_id}")
def get_customer(customer_id: int, db: Session = Depends(get_db)):
    c = db.query(Customer).filter(Customer.id == customer_id).first()
    if not c:
        raise HTTPException(status_code=404, detail="Customer not found")
    return _serialize(c)


@router.post("/")
def create_customer(payload: CustomerCreate, db: Session = Depends(get_db)):
    name = payload.name.strip()
    email = payload.email.strip()
    mobile = payload.mobile.strip()

    if not name or not email or not mobile:
        raise HTTPException(status_code=400, detail="Name, email, and mobile are required")

    existing = db.query(Customer).filter((Customer.mobile == mobile) | (Customer.email == email)).first()
    if existing:
        raise HTTPException(status_code=400, detail="Customer with this mobile or email already exists")

    c = Customer(
        name=name,
        email=email,
        mobile=mobile,
        uniqueCustomerId=mobile,
        status="not-started",
    )
    db.add(c)
    db.commit()
    db.refresh(c)
    return _serialize(c)


@router.put("/{customer_id}")
def update_customer(customer_id: int, payload: CustomerUpdate, db: Session = Depends(get_db)):
    c = db.query(Customer).filter(Customer.id == customer_id).first()
    if not c:
        raise HTTPException(status_code=404, detail="Customer not found")

    name = payload.name.strip()
    email = payload.email.strip()
    mobile = payload.mobile.strip()

    if not name or not email or not mobile:
        raise HTTPException(status_code=400, detail="Name, email, and mobile are required")

    conflict = db.query(Customer).filter(
        ((Customer.mobile == mobile) | (Customer.email == email)),
        Customer.id != customer_id
    ).first()
    if conflict:
        raise HTTPException(status_code=400, detail="Another customer with this mobile or email already exists")

    c.name = name
    c.email = email
    c.mobile = mobile
    db.add(c)
    db.commit()
    db.refresh(c)
    return _serialize(c)


@router.put("/{customer_id}/assign-agent")
def assign_agent(customer_id: int, payload: AssignAgentPayload, db: Session = Depends(get_db)):
    c = db.query(Customer).filter(Customer.id == customer_id).first()
    if not c:
        raise HTTPException(status_code=404, detail="Customer not found")

    if payload.agentId is not None:
        agent = db.query(Agent).filter(Agent.id == payload.agentId).first()
        if not agent:
            raise HTTPException(status_code=404, detail="Agent not found")
        c.agentId = payload.agentId
    else:
        c.agentId = None

    db.add(c)
    db.commit()
    db.refresh(c)
    return _serialize(c)


@router.delete("/{customer_id}")
def delete_customer(customer_id: int, db: Session = Depends(get_db)):
    c = db.query(Customer).filter(Customer.id == customer_id).first()
    if not c:
        raise HTTPException(status_code=404, detail="Customer not found")
    c.isActive = False
    db.add(c)
    db.commit()
    return {"status": "ok"}

File: DSA-loan-management/dsa-mgmt-fe/src/services/customers.ts
import axios from 'axios'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://127.0.0.1:8000'

export type CustomerInput = {
  name: string
  email: string
  mobile: string
}

export type Customer = {
  id: number
  name: string
  email: string
  mobile: string
  uniqueCustomerId?: string
  agentId?: number | null
  agentName?: string | null
  agentPhoto?: string | null
  bankId?: number | null
  bankName?: string | null
  bankLogo?: string | null
  status: string
  description?: string | null
}

export async function fetchCustomers(agentId?: number): Promise<Customer[]> {
  const params = agentId !== undefined && agentId !== null ? { agent_id: agentId } : {}
  const res = await axios.get(`${API_BASE_URL}/api/customers`, { params })
  return res.data || []
}

export async function assignCustomerAgent(customerId: number, agentId: number | null): Promise<Customer> {
  const res = await axios.put(`${API_BASE_URL}/api/customers/${customerId}/assign-agent`, { agentId })
  return res.data
}

export async function updateCustomerStatus(
  customerId: number,
  payload: { status: string; bankId?: number | null; description?: string | null }
): Promise<Customer> {
  const res = await axios.put(`${API_BASE_URL}/api/customers/${customerId}/status`, payload)
  return res.data
}

export async function createCustomer(payload: CustomerInput): Promise<Customer> {
  const res = await axios.post(`${API_BASE_URL}/api/customers`, payload)
  return res.data
}

export async function updateCustomer(id: number, payload: CustomerInput): Promise<Customer> {
  const res = await axios.put(`${API_BASE_URL}/api/customers/${id}`, payload)
  return res.data
}

export async function deleteCustomer(id: number): Promise<{ status: string }> {
  const res = await axios.delete(`${API_BASE_URL}/api/customers/${id}`)
  return res.data
}

export async function addCustomer(payload: CustomerInput) {
  const res = await axios.post(`${API_BASE_URL}/api/auth/customer/add`, payload)
  return res.data
}